pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,493 @@
+		return true
+	} else if err := c.syncConsistenHash(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Syncing failed, requeue", "eip", key)
+	}
+	return true
+}
+
+func (c *Cluster) syncConsistenHash(eipName string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).InfoS("Finished syncing consistentHash", "eip", eipName, "durationTime", time.Since(startTime))
+	}()
+
+	c.consistentHashRWMutex.Lock()
+	defer c.consistentHashRWMutex.Unlock()
+	eip, err := c.externalIPPoolLister.Get(eipName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			delete(c.consistentHashMap, eipName)
+			return nil
+		}
+		return err
+	}
+
+	if err := c.syncConsistentHashByEIP(eip); err != nil {
+		klog.ErrorS(err, "Sync consistentHash by EIP failed", "EIP", eip.Name)
+		return err
+	}
+
+	return nil
+}
+
+func newNodeConsistentHashMap() *consistenthash.Map {
+	return consistenthash.New(defaultVirtualNodeReplicas, defaultHashFn)
+}
+
+// syncConsistentHashByEIP refresh the consistentHashMap
+func (c *Cluster) syncConsistentHashByEIP(eip *v1alpha2.ExternalIPPool) error {
+	nodeSel, err := metav1.LabelSelectorAsSelector(&eip.Spec.NodeSelector)
+	if err != nil {
+		return err
+	}
+	nodes, err := c.nodeLister.List(nodeSel)
+	if err != nil {
+		return err
+	}
+	c.consistentHashMap[eip.Name] = newNodeConsistentHashMap()
+	aliveNodes := c.aliveNodeList()
+	// node alive and node labels matches with externalIPPool nodeSelector
+	aliveAndMatchedNodes := sets.NewString()
+	for _, node := range nodes {
+		nodeName := node.Name
+		if aliveNodes.Has(nodeName) {
+			aliveAndMatchedNodes.Insert(nodeName)
+		}
+	}
+	c.consistentHashMap[eip.Name].Add(aliveAndMatchedNodes.List()...)
+	c.notify(eip.Name)
+	return nil
+}
+
+func (c *Cluster) handleClusterNodeEvents(nodeEvent *memberlist.NodeEvent) {
+	eips, err := c.externalIPPoolLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Handle clusterNode events failed")
+	}
+	switch node, event := nodeEvent.Node, nodeEvent.Event; event {
+	case memberlist.NodeJoin:
+		// when node join cluster, all matched externalIPPools consistentHash should changed
+		node, err := c.nodeLister.Get(node.Name)
+		if err != nil {
+			klog.ErrorS(err, "Processing clusterNode events failed")
+			return
+		}
+		count := 0
+		for _, eip := range eips {
+			if c.matchNodeSelector(node, eip) {
+				c.queue.Add(eip.Name)
+				klog.V(5).InfoS("Node join event", "nodeName", node.String(), "eip", eip.Name)
+				count++
+			}
+		}
+		klog.InfoS("Node event: join node", "nodeName", node.Name, "eipNum", count)
+	case memberlist.NodeLeave:
+		// when node leave cluster, the node may failed or deleted, all externalIPPools consistentHash maybe changed
+		for _, eip := range eips {
```

Comment:
I think we only need to handle the failed case if we handle the node DELETE event in the event handler as if the node still exists, it must be a failed case and we can figure out affected pools by matching the labels, if the node doesn't exist, we don't need to process it as affected pools must have been enqueued by DELETE event handler.

Reply:
done